### ISSUE TITLE
Configured-out generator blocks are still parsed and cause errors

### ISSUE THREAD

**Issue by gkgoat1** [rust-lang/rust#152866]
I tried this code (from https://github.com/portal-co/wax a003a01d3e07237a6f98ebd695e7b343b08d203f):

```rust
#[cfg(feature = "gen-blocks")]
impl<Context, E> OperatorIterSource<Context, E> for Tuple {
    for_tuples!(where #(Tuple: OperatorIterSource<Context, E>)*);
    fn operators<'a>(
        &'a self,
        ctx: &'a mut Context,
    ) -> Box<dyn Iterator<Item = Result<Operator<'static>, E>> + 'a>
    where
        E: 'a,
    {
        Box::new(gen move {
            for_tuples!(#(for op in Tuple.operators(ctx){
                    yield op;
                });*)
        })
    }
}
``` 
(gen-blocks is a disabled feature)

I expected to see this happen: No error

Instead, this happened: An error, ```gen blocks are experimental
see issue #117078 <https://github.com/rust-lang/rust/issues/117078>```, despite the code being configured out and therefore shouldn't be parsed

### Meta

`rustc --version --verbose`:
```
rustc 1.93.1 (01f6ddf75 2026-02-11)
binary: rustc
commit-hash: 01f6ddf7588f42ae2d7eb0a2f21d44e8e96674cf
commit-date: 2026-02-11
host: aarch64-apple-darwin
release: 1.93.1
LLVM version: 21.1.8
```
Fails identically on beta, nightly, and with both `rustc` and `rust-analyzer`

<details><summary>Backtrace</summary>
<p>

No backtrace, it's just a normally reported error

</p>
</details>

**Comment by fmease** [comment]
> despite the code being configured out and therefore shouldn't be parsed

Just to clear up a miscomception, things that are cfg'ed out are still being parsed pre-expansion.

In any case, what you're witnessing is called a pre-expansion feature gate and it works precisely as intended here. All modern unstable syntactic features are gated via this mechanism, only decade-old features are protected by legacy post-expansion feature gates.

The latter led to a bunch of widely used crates in rhe ecosystem adopting the "`#[cfg(feature = "nightly")]` pattern" where the crate could be used (primarily) by stable users with the option for nightly users to enable extra functionality that depends on unstable features. What does that mean for rustc devs changing or removing the unstable syntax? Well, it would break tons of stable users which is less than ideal.

**Comment by fmease** [comment]
The workaround for users is to wrap any uses of unstable syntax in a (declarative) macro which you only invoke if the cfg is active. The body of decl macros obviously won't be parsed ahead of time that's why it works.
